### ISSUE TITLE
Rewrite HTTP client

### ISSUE THREAD

**Issue by lvillani** [just-install/just-install#287]
Build a new set of HTTP client helper functions built on top of Go's http/client but written with the following principles in mind:

* Handle errors correctly instead of using log.Fatal everywhere
* Better layering of helper functions (currently it's a bodged-together mess)
* Add support for features we'll need to implement registry v5 (cookie bag, etc)
* Handle partial downloads: at the moment a partial download results in trying to execute a corrupted executable the next time we try to install a package.
* Proper local file names: we run a hashing function on the URL and use that to derive a local file name. We should be more like browsers and just figure out how to get the real name of the file that we are downloading. This also makes it easier to debug problems when we open `%TEMP%\just-install` in a file manager.

TL; DR: we need something like `wget` to download installers, and something more flexible to handle edge cases, link checks, etc that is written in pure Go. Due to the amount of stuff that we have to support is kinda hard to design a coherent API.

**Comment by lvillani** [comment]
We already support HTTPS downloads. We are not going to force them like e.g. HTTPS Everywhere does in browsers, though.
